php_phongo.c

```diff
@@ -2841,18 +2841,379 @@ static bool phongo_manager_set_auto_encryption_opts(php_phongo_manager_t* manage
 	return retval;
 }
 /* }}} */
+
+static mongoc_client_encryption_opts_t* phongo_clientencryption_opts_from_zval(mongoc_client_t* client, zval* options TSRMLS_DC) /* {{{ */
+{
+	mongoc_client_encryption_opts_t* opts;
+
+	opts = mongoc_client_encryption_opts_new();
+
+	if (!options || Z_TYPE_P(options) != IS_ARRAY) {
+		return opts;
+	}
+
+	if (php_array_existsc(options, "keyVaultClient")) {
+		zval* key_vault_client = php_array_fetch(options, "keyVaultClient");
+
+		if (Z_TYPE_P(key_vault_client) != IS_OBJECT || !instanceof_function(Z_OBJCE_P(key_vault_client), php_phongo_manager_ce TSRMLS_CC)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"keyVaultClient\" encryption option to be %s, %s given", ZSTR_VAL(php_phongo_manager_ce->name), PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(key_vault_client));
+			goto cleanup;
+		}
+
+		mongoc_client_encryption_opts_set_keyvault_client(opts, Z_MANAGER_OBJ_P(key_vault_client)->client);
+	} else {
+		mongoc_client_encryption_opts_set_keyvault_client(opts, client);
+	}
+
+	if (php_array_existsc(options, "keyVaultNamespace")) {
+		char*     keyvault_namespace;
+		char*     db_name;
+		char*     coll_name;
+		int       plen;
+		zend_bool pfree;
+
+		keyvault_namespace = php_array_fetchc_string(options, "keyVaultNamespace", &plen, &pfree);
+
+		if (!phongo_split_namespace(keyvault_namespace, &db_name, &coll_name)) {
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected \"keyVaultNamespace\" encryption option to contain a full collection name");
+
+			goto cleanup;
```

Comment:
If we end up jumping to `cleanup`, might we still need to conditionally call `efree(keyvault_namespace)` if `pfree` is `true`?

I checked, and you did handle that case in `phongo_manager_set_auto_encryption_opts`. Perhaps this was written before I caught that in #1071, though. Also, is there a particular reason you used `str_efree` in the auto opts, but `efree` here? Our own `phongo_compat.h` defines it as the same for PHP 7, but the [original function for PHP 5.x](http://lxr.jpauli.tech/source/xref/php-5.6/Zend/zend_string.h#53) has a check for interned strings which might be important. It looks like we use `str_efree` a few other places in the driver in a similar fashion, so that may be correct but I couldn't tell you why off the top of my head.

While looking into this, I think I found a few places where we might just be using `efree()` on something returned from `php_array_fetch_string`. `php_phongo_fetch_ssl_opt_string` may be one instance where we tend to use `str_efree` but may use `efree` directly at least once. Would you mind opening a ticket to investigate this? It's likely we won't even get to it if we drop 5.6 support first, but if so we could always close it accordingly.

Reply:
I changed handling of `keyvault_namespace` to be equivalent to the handling in `phongo_manager_set_auto_encryption_opts`. I've also created [PHPC-1527](https://jira.mongodb.org/browse/PHPC-1527) to investigate other instances where we may be using `efree` instead of `str_efree`.